### PR TITLE
Do not use async when invoking code actions

### PR DESCRIPTION
**Pull request by jmarolf**
**Customer scenario**

Applying any codefix will never show a wait dialog, no matter how long the operation takes.  This includes fixes in a single application of a fix, or fixall in Document, Project, or Solution.

**Bugs this fixes:** 

https://github.com/dotnet/roslyn/issues/15411
https://github.com/dotnet/roslyn/issues/15761
https://github.com/dotnet/roslyn/issues/15775

**Workarounds, if any**

None

**Risk**

Low, we are reverting to earlier behavior where we do not use async when invoking code actions.

**Performance impact**

Low, we are reverting to earlier behavior where we do not use async when invoking code actions.

**Is this a regression from a previous update?**

Yes, this was introduced in RC.2

**Root cause analysis:**

Cyrus made a large refactoring to Fixall in RC.2.  All of our internal understanding of how the threaded wait dialog works led us to believe that creation on a forground thread and then awaiting a background thread was a legal thing to do.  Turns out creation on a forground thread then yielding to a background thread causes the threaded wait dialog to never pop up.  The fix is to not use async code in combination with the threaded wait dialog.

**How was the bug found?**

dogfooding

### CODE REVIEW

**Comment by jmarolf** [comment]
@dotnet/roslyn-ide @CyrusNajmabadi please review

**Comment by Pilchie** [comment]
I'm not convinced this is right.  We create the TWD on the foreground in just about every other feature.  It was specifically designed to allow putting up UI while doing blocking operations on the foreground thread.

Without more understanding, I'm not in favor of this change.

**Comment by jmarolf** [comment]
@Pilchie Sorry, I've updated my description to include an important detaill. we aren't blocking operations on the forground thread. Everywhere else in roslyn we call the [wait extension method
](http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Host/IWaitIndicator.cs,dc5c35ae8c15ae98,references).  Everwhere else the TWD is created and then the action is done on the same thread.  Here we create TWD on the forground thread and then await a background operation.

File: src/EditorFeatures/Core/Implementation/Suggestions/SuggestedAction.cs [hunk 1]
```diff
@@ -102,44 +102,45 @@ public void Invoke(CancellationToken cancellationToken)
...
             // Always wrap whatever we're doing in a threaded wait dialog.
-            using (var context = this.SourceProvider.WaitIndicator.StartWait(CodeAction.Title, CodeAction.Message, allowCancel: true, showProgress: true))
-            using (var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, context.CancellationToken))
+            using (var waitContext = await GetWaitContextAsync(cancellationToken).ConfigureAwait(true))
+            using (var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, waitContext.CancellationToken))
             {
-                // Yield the UI thread so that the light bulb can be dismissed.  This is necessary
-                // as some code actions may be long running, and we don't want the light bulb to
-                // stay on screen.
-                await Task.Yield();
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/Suggestions/SuggestedAction.cs]: Why was this removed?

**Comment by CyrusNajmabadi** [comment]
> All of our internal understanding of how the threaded wait dialog works led us to believe that creation on a forground thread and then awaiting a background thread was a legal thing to do. Turns out creation on a forground thread then yielding to a background thread causes the threaded wait dialog to never pop up

Er... this violates my understanding of the TWD.  And it is not how the TWD works in any other context.  I believe we *always* create the TWD on hte UI thread.

**Comment by CyrusNajmabadi** [comment]
> Everywhere else in roslyn we call the wait extension method

As an experiment, can you try actually 'Wait'ing on the async work executed inside the TWD block?

File: src/EditorFeatures/Core/Implementation/Suggestions/SuggestedAction.cs [hunk 2]
```diff
@@ -102,44 +102,45 @@ public void Invoke(CancellationToken cancellationToken)
...
-
-                this.AssertIsForeground();
+                AssertIsForeground();
 
                 // Then proceed and actually do the invoke.
-                await InvokeAsync(context.ProgressTracker, linkedSource.Token).ConfigureAwait(true);
+                await InvokeAsync(waitContext.ProgressTracker, linkedSource.Token).ConfigureAwait(true);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/Suggestions/SuggestedAction.cs]: i.e. write: ```InvokeAsync(waitContext.ProgressTracker, linkedSource.Token).Wait(...)```.

Instead of calling StartWait, you can also just call ```WaitIndicator.Wait(...)``` and pass it a synchronous lambda.

Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/Suggestions/SuggestedAction.cs]: Actually, that may not work at all.  We'll be blocking on the UI thread.  But many tasks we kick off end up await'ing work happening on the UI thread. So that will probably just deadlock.  Regardless, we need to understand why on earth popping on the TWD on the FG thread isn't actually functioning :(

**Comment by CyrusNajmabadi** [comment]
You can contact JeffRo about the TWD.

**Comment by jmarolf** [comment]
Yeah, I'm going to step through the TWD sources to get a better explaination

**Comment by jmarolf** [comment]
TWD cannot be created on the UI thread and work correctly with an await.  If you create the TWD on the UI thread you need to block the UI.  We cannot block the UI thread at SuggestedAction as it is possible to deadlock.  If we create the TWD on a background thread it will switch to the UI thread automatically for creation and then switch back to the calling thread.  In this second case we will not block the UI thread while a codefix is applied, but the wait dialog will popup and work correctly.  In this second case the user can still type in the editor while the fix is being applied.  If the user types anything the codefix application will fail (which looks like applying the codefix did nothing).  At this point we can either accept this PR or revert Cyrus' changes.

**Comment by jmarolf** [comment]
Updated PR so it now does the following:
1. Create TWD on background thread so avoid deadlock and can update progress on TWD correctly
2. Block application of codefixes so user cannot interact with editor while they are being applied
3. Introduce Try/catch for update call for race condition where signal to cancel dialog is processed after dialog has been destroyed.

I would like someone to pull down this branch and do some independent testing before this is approved.  Preferable @mavasani as he has familiarity with the original implementation.

**Comment by jmarolf** [comment]
test windows_eta_open_prtest please

File: src/EditorFeatures/Core/Implementation/Suggestions/SuggestedAction.cs [hunk 3]
```diff
@@ -102,44 +102,45 @@ public void Invoke(CancellationToken cancellationToken)
 
         private async Task YieldThenInvokeAsync(CancellationToken cancellationToken)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/Suggestions/SuggestedAction.cs]: Rename since we don't actually yield.

Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/Suggestions/SuggestedAction.cs]: This method is sorely needing some comments.  Especially why we do a blocking wait in an async method.

File: src/EditorFeatures/Core/Implementation/Suggestions/SuggestedAction.cs [hunk 4]
```diff
@@ -102,44 +102,45 @@ public void Invoke(CancellationToken cancellationToken)
...
-                this.AssertIsForeground();
+                AssertIsForeground();
 
                 // Then proceed and actually do the invoke.
-                await InvokeAsync(context.ProgressTracker, linkedSource.Token).ConfigureAwait(true);
+                InvokeAsync(waitContext.ProgressTracker, linkedSource.Token).Wait(linkedSource.Token);
             }
         }
 
+        private Task<IWaitContext> GetWaitContextAsync(CancellationToken cancellationToken)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/Suggestions/SuggestedAction.cs]: Explain this.  Also, just inline as well.

File: src/EditorFeatures/Core/Implementation/Suggestions/SuggestedAction.cs [hunk 5]
```diff
@@ -149,30 +150,28 @@ public void Invoke(CancellationToken cancellationToken)
...
-            if (actionWithOptions != null)
+            if (CodeAction is CodeActionWithOptions actionWithOptions)
             {
                 var options = actionWithOptions.GetOptions(cancellationToken);
                 if (options != null)
                 {
                     // ConfigureAwait(true) so we come back to the same thread as 
                     // we do all application on the UI thread.
                     operations = await GetOperationsAsync(actionWithOptions, options, cancellationToken).ConfigureAwait(true);
-                    this.AssertIsForeground();
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/Suggestions/SuggestedAction.cs]: since we ConfigureAwait(true) here, how do we avoid deadlocking since you called 'Wait' above?

File: src/EditorFeatures/Core/Implementation/Suggestions/SuggestedAction.cs [hunk 6]
```diff
@@ -102,44 +102,45 @@ public void Invoke(CancellationToken cancellationToken)
 
         private async Task YieldThenInvokeAsync(CancellationToken cancellationToken)
         {
-            this.AssertIsForeground();
+            AssertIsForeground();
 
             // Always wrap whatever we're doing in a threaded wait dialog.
-            using (var context = this.SourceProvider.WaitIndicator.StartWait(CodeAction.Title, CodeAction.Message, allowCancel: true, showProgress: true))
-            using (var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, context.CancellationToken))
+            using (var waitContext = await GetWaitContextAsync(cancellationToken).ConfigureAwait(true))
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/Suggestions/SuggestedAction.cs]: comment why ConfigureAwait(true) is appropriate.

File: src/EditorFeatures/Core/Implementation/Suggestions/SuggestedAction.cs [hunk 7]
```diff
@@ -102,44 +102,45 @@ public void Invoke(CancellationToken cancellationToken)
...
-                // as some code actions may be long running, and we don't want the light bulb to
-                // stay on screen.
-                await Task.Yield();
-
-                this.AssertIsForeground();
+                AssertIsForeground();
 
                 // Then proceed and actually do the invoke.
-                await InvokeAsync(context.ProgressTracker, linkedSource.Token).ConfigureAwait(true);
+                InvokeAsync(waitContext.ProgressTracker, linkedSource.Token).Wait(linkedSource.Token);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/Suggestions/SuggestedAction.cs]: instead of doing the blocking wait here, can we instead move the 'Wait' up to 'Invoke'?  That way 'Invoke' stays synchronous and prevents users from interacting with the UI.  On the plus side, we also won't have any blocking in these async methods.

Note: we'll need a try/catch since Wait(...) can then throw cancellation exceptions.

**Comment by jmarolf** [comment]
After trying out more combinations.  I've discovered that this too can deadlock the UI thread.  I'm going to remove all asynchronicity from SuggestedAction as it is fundamentally incompatible with TWD. If we want to use async we will need to write our own blocking dialog.

**Comment by jmarolf** [comment]
ready for review @CyrusNajmabadi @jasonmalinowski @Pilchie @dotnet/roslyn-ide

File: src/EditorFeatures/Core/Implementation/Suggestions/SuggestedAction.cs [hunk 8]
```diff
@@ -90,89 +90,57 @@ protected Task<ImmutableArray<CodeActionOperation>> GetPreviewOperationsAsync(Ca
...
-        }
-
-        private async Task YieldThenInvokeAsync(CancellationToken cancellationToken)
-        {
-            this.AssertIsForeground();
-
-            // Always wrap whatever we're doing in a threaded wait dialog.
-            using (var context = this.SourceProvider.WaitIndicator.StartWait(CodeAction.Title, CodeAction.Message, allowCancel: true, showProgress: true))
-            using (var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, context.CancellationToken))
+            var waitResult = this.SourceProvider.WaitIndicator.Wait(CodeAction.Title, CodeAction.Message, allowCancel: true, showProgress: true, action: waitContext =>
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/Suggestions/SuggestedAction.cs]: you don't need/use waitResult.

File: src/EditorFeatures/Core/Implementation/Suggestions/SuggestedAction.cs [hunk 9]
```diff
@@ -90,89 +90,57 @@ protected Task<ImmutableArray<CodeActionOperation>> GetPreviewOperationsAsync(Ca
...
             {
                 var options = actionWithOptions.GetOptions(cancellationToken);
                 if (options != null)
                 {
-                    // ConfigureAwait(true) so we come back to the same thread as 
-                    // we do all application on the UI thread.
-                    operations = await GetOperationsAsync(actionWithOptions, options, cancellationToken).ConfigureAwait(true);
-                    this.AssertIsForeground();
+                    operations = GetOperationsAsync(actionWithOptions, options, cancellationToken).WaitAndGetResult(cancellationToken);
+                    AssertIsForeground();
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/Suggestions/SuggestedAction.cs]: could be removed at this poing.

**Comment by CyrusNajmabadi** [comment]
👍 Seems like a really good solution given our current constraints.  Thanks!

**Comment by CyrusNajmabadi** [comment]
My only recommendation is that we comment the code we ahve to explain the reasoning behind things.

**Comment by jmarolf** [comment]
review please @jasonmalinowski @Pilchie @dotnet/roslyn-ide

**Comment by jmarolf** [comment]
test windows_release_unit64_prtest please

**Comment by jmarolf** [comment]
@srivatsn I've gotten buddy testing from @tannergooding and feel confident in merging this in.

**Comment by tannergooding** [comment]
👍

**Comment by CyrusNajmabadi** [comment]
Awesome!
